addons/account_edi_ubl/models/account_edi_format.py

```diff
@@ -16,6 +16,22 @@
 class AccountEdiFormat(models.Model):
     _inherit = 'account.edi.format'
 
+    def _create_invoice_from_ubl(self, tree):
+        invoice = self.env['account.move']
+        journal = invoice._context.get('default_journal_id')
+        if journal:
+            move_type = 'out_invoice' if journal.type == 'sale' else 'in_invoice'
+        else:
+            move_type = invoice._context.get('default_move_type', 'in_invoice')
+            journal = invoice.with_context(default_move_type=move_type)._get_default_journal()
+
+        elements = tree.find('.//{*}InvoiceTypeCode')
+        if elements and elements[0].text == '381':
+            move_type = 'in_refund' if move_type == 'in_invoice' else 'out_refund'
+
+        invoice = invoice.with_context(default_move_type=move_type, default_journal_id=journal.id)
+        return self._import_ubl(tree, invoice)
```

Comment:
```
elements = tree.xpath('//cbc:InvoiceTypeCode', namespaces=namespaces)
        if elements:
            type_code = elements[0].text
            move_type = 'in_refund' if type_code == '381' else 'in_invoice'
        else:
            move_type = 'in_invoice'
```
... is giving you the move type. Add it into the context.

When creating a new invoice from scratch:
`self.env['account.move']._get_default_journal()`
...is already managing the context to retrieve the right account.

When updating an existing invoice:
you already have the move_type and the journal, just need to add them into the context. However, you can change dynamicaly the move_type so you need to ensure the move_type in the tree is the same as the one set on the move.